File: src/config/config_manager.nim
import os, tables, strutils
import core/session_state
import core/evaluator
import parser/parser

proc directoryExists(path: string): bool =
  dirExists(path)

proc getFileDirectory*(homeDir: string): string =
  let main = homeDir & ".config/swell/"
  let backup = homeDir & ".swell/"
  let fileName = "config.swell"
  let backupFile = backup & fileName
  let mainFile = main & fileName

  if directoryExists(main) and fileExists(mainFile):
    return mainFile
  elif directoryExists(backup) and fileExists(backupFile):
    return backupFile
  else:
    if not directoryExists(main):
      try:
        createDir(main)
      except OSError:
        echo "Failed to create directory for file."
    try:
      writeFile(mainFile, "echo Hello World")
    except IOError:
      echo "Failed to initialise configuration file."
    return mainFile

proc loadConfiguration*(state: SessionState) =
  let fileDir = getFileDirectory(state.homeDir)
  echo "debug filedir: ", fileDir
  var configFileContents = ""
  try:
    configFileContents = readFile(fileDir)
  except IOError:
    echo "error reading configuration file, must be utf8 encoded!"
  if configFileContents.len > 0:
    let parsedConfigFile = configFileContents.split('\n')
    for l in parsedConfigFile:
      if l.strip().len > 0:
        let node = parse(l)
        evaluate(node, state)

proc updateAvailableCommands*(state: SessionState) =
  var commands: seq[string] = @[]
  let pathEnv = getEnv("PATH", "")
  let pathDirectories = pathEnv.split(':')
  for dir in pathDirectories:
    if not dirExists(dir):
      continue
    for kind, path in walkDir(dir):
      if kind == pcFile or kind == pcLinkToFile:
        let (_, name, _) = splitFile(path)
        commands.add(name)
  state.availableCommands = commands

proc initSessionState*(): SessionState =
  let homeDir = getHomeDir()
  let currentDir = getCurrentDir()
  let env = {"TERM": "xterm-256color", "COLORTERM": "truecolor"}.toTable
  var newSession = newSessionState(env, homeDir, currentDir)
  updateAvailableCommands(newSession)
  return newSession

File: src/config/config_manager.rs
use std::collections::HashMap;
use std::env;
use std::fs;
use std::path::Path;

use crate::core::session_state::SessionState;
use crate::core::evaluator::evaluate;
use crate::parser::parser::parse;

pub struct ConfigManager;

impl ConfigManager {
    pub fn get_file_directory(home_dir: &str) -> String {
        let main_dir = format!("{}/.config/swell/", home_dir);
        let backup_dir = format!("{}/.swell/", home_dir);
        let file_name = "config.swell";
        let main_file = format!("{}{}", main_dir, file_name);
        let backup_file = format!("{}{}", backup_dir, file_name);

        if Path::new(&main_dir).is_dir() && Path::new(&main_file).is_file() {
            return main_file;
        } else if Path::new(&backup_dir).is_dir() && Path::new(&backup_file).is_file() {
            return backup_file;
        }

        // Create the main directory if it doesn't exist
        if !Path::new(&main_dir).is_dir() {
            if let Err(e) = fs::create_dir_all(&main_dir) {
                eprintln!("Failed to create directory for file: {}", e);
            }
        }

        // Write a default config
        if let Err(e) = fs::write(&main_file, "echo Hello World") {
            eprintln!("Failed to initialise configuration file: {}", e);
        }

        main_file
    }

    pub fn load_configuration(state: &mut SessionState) {
        let file_dir = Self::get_file_directory(&state.home_dir);
        println!("debug filedir: {}", file_dir);

        match fs::read_to_string(&file_dir) {
            Ok(contents) => {
                if !contents.is_empty() {
                    for line in contents.lines() {
                        let node = parse(line);
                        evaluate(node, state);
                    }
                }
            }
            Err(_) => {
                eprintln!("error reading configuration file, must be utf8 encoded!");
            }
        }
    }

    pub fn init_session_state() -> SessionState {
        let home_dir = env::var("HOME").unwrap_or_else(|_| String::from("/"));
        let current_dir = env::current_dir()
            .map(|p| p.to_string_lossy().to_string())
            .unwrap_or_else(|_| String::from("/"));

        let mut environment = HashMap::new();
        environment.insert("TERM".to_string(), "xterm-256color".to_string());
        environment.insert("COLORTERM".to_string(), "truecolor".to_string());

        let mut session = SessionState::new(environment, home_dir, current_dir);
        Self::update_available_commands(&mut session);
        session
    }

    pub fn update_available_commands(state: &mut SessionState) {
        let mut commands: Vec<String> = Vec::new();
        let path = env::var("PATH").unwrap_or_default();
        let path_directories: Vec<&str> = path.split(':').collect();

        for dir in path_directories {
            if let Ok(entries) = fs::read_dir(dir) {
                for entry in entries.flatten() {
                    if let Some(name) = entry.file_name().to_str() {
                        commands.push(name.to_string());
                    }
                }
            }
        }

        state.available_commands = commands;
    }
}
